src/main/java/smartthings/konsumer/filterchain/MessageFilterChain.java

```diff
@@ -10,37 +10,37 @@
 
 import java.util.*;
 
-public class MessageFilterChain {
+public class MessageFilterChain<K, V, R> {
 	private final static Logger log = LoggerFactory.getLogger(MessageFilterChain.class);
 
-	private final MessageProcessor processor;
-	private final List<MessageFilter> filters;
+	private final MessageProcessor<K, V, R> processor;
+	private final List<MessageFilter<K, V, R>> filters;
 
-	public MessageFilterChain(final MessageProcessor processor, final MessageFilter... messageFilters) {
+	public MessageFilterChain(final MessageProcessor<K, V, R> processor, final List<MessageFilter<K, V, R>> messageFilters) {
 		this.processor = processor;
-		this.filters = Collections.unmodifiableList(new ArrayList<MessageFilter>(Arrays.asList(messageFilters)));
+		this.filters = Collections.unmodifiableList(new ArrayList<>(messageFilters)); //make defensive copy
 	}
 
-	public void handle(MessageAndMetadata<byte[], byte[]> originalMessageAndMetadata,
+	public R handle(MessageAndMetadata<K, V> originalMessageAndMetadata,
```

Comment:
Out of curiosity what use case caused you to want this? I originally thought of this as kind of like a servlet filter style pattern, but I can see some value in this.  Always could use Void if you don't want R to have a value.